File: sdk-rust/src/permission_cache.rs
//! Pluggable cache for the per-identity effective-permission set.

use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Mutex;
use std::time::{Duration, Instant};

use crate::types::{EffectivePermission, PermissionCacheKey};

/// Sits in front of the recursive-CTE permission-resolution query.
///
/// The library ships an [`InMemoryPermissionCache`] default. Hosts that
/// need a shared cache (Redis, Memcached, etc.) implement this trait
/// against their cache client and pass it to
/// `GatedhouseConfigBuilder::permission_cache(...)`.
///
/// Implementations must be `Send + Sync` and safe for concurrent use.
pub trait PermissionCache: Send + Sync {
    /// Return the cached list for `(identity_id, org_id)`, or `None` on
    /// miss / expiry.
    fn get(&self, identity_id: &str, org_id: &str) -> Option<Vec<EffectivePermission>>;

    /// Cache the effective-permission list. Implementations should
    /// treat the supplied vec as immutable (clone if needed).
    fn put(&self, identity_id: &str, org_id: &str, permissions: Vec<EffectivePermission>);

    /// Drop the cache entry for one identity in one org.
    fn invalidate(&self, identity_id: &str, org_id: &str);

    /// Drop every cached entry.
    fn invalidate_all(&self);
}

/// Default [`PermissionCache`]: thread-safe, in-process, TTL-expiring
/// `HashMap`. Lazy eviction on read.
///
/// Exposes counters for hits, misses, puts, and invalidations as
/// runtime observability — and to let tests prove caching behaviour,
/// not just outcomes.
pub struct InMemoryPermissionCache {
    ttl: Duration,
    entries: Mutex<HashMap<PermissionCacheKey, Entry>>,
    hits: AtomicU64,
    misses: AtomicU64,
    puts: AtomicU64,
    targeted_invalidations: AtomicU64,
    wholesale_invalidations: AtomicU64,
}

struct Entry {
    permissions: Vec<EffectivePermission>,
    expires_at: Instant,
}

impl InMemoryPermissionCache {
    /// Conservative default — covers brief bursts of activity for an
    /// identity while keeping the window for stale data after an
    /// out-of-band schema edit short.
    pub const DEFAULT_TTL: Duration = Duration::from_secs(60);

    pub fn new() -> Self {
        Self::with_ttl(Self::DEFAULT_TTL)
    }

    pub fn with_ttl(ttl: Duration) -> Self {
        Self {
            ttl,
            entries: Mutex::new(HashMap::new()),
            hits: AtomicU64::new(0),
            misses: AtomicU64::new(0),
            puts: AtomicU64::new(0),
            targeted_invalidations: AtomicU64::new(0),
            wholesale_invalidations: AtomicU64::new(0),
        }
    }

    // ---- observability ----------------------------------------------------

    pub fn hit_count(&self) -> u64 {
        self.hits.load(Ordering::Relaxed)
    }

    pub fn miss_count(&self) -> u64 {
        self.misses.load(Ordering::Relaxed)
    }

    pub fn put_count(&self) -> u64 {
        self.puts.load(Ordering::Relaxed)
    }

    pub fn targeted_invalidation_count(&self) -> u64 {
        self.targeted_invalidations.load(Ordering::Relaxed)
    }

    pub fn wholesale_invalidation_count(&self) -> u64 {
        self.wholesale_invalidations.load(Ordering::Relaxed)
    }

    pub fn size(&self) -> usize {
        self.entries.lock().expect("cache lock poisoned").len()
    }

    /// Reset all counters to zero. Does not touch cached entries.
    pub fn reset_stats(&self) {
        self.hits.store(0, Ordering::Relaxed);
        self.misses.store(0, Ordering::Relaxed);
        self.puts.store(0, Ordering::Relaxed);
        self.targeted_invalidations.store(0, Ordering::Relaxed);
        self.wholesale_invalidations.store(0, Ordering::Relaxed);
    }
}

impl Default for InMemoryPermissionCache {
    fn default() -> Self {
        Self::new()
    }
}

impl PermissionCache for InMemoryPermissionCache {
    fn get(&self, identity_id: &str, org_id: &str) -> Option<Vec<EffectivePermission>> {
        let key = PermissionCacheKey::new(identity_id, org_id);
        let now = Instant::now();
        let mut entries = self.entries.lock().expect("cache lock poisoned");
        match entries.get(&key) {
            Some(entry) if entry.expires_at > now => {
                let result = entry.permissions.clone();
                drop(entries);
                self.hits.fetch_add(1, Ordering::Relaxed);
                Some(result)
            }
            Some(_) => {
                entries.remove(&key);
                drop(entries);
                self.misses.fetch_add(1, Ordering::Relaxed);
                None
            }
            None => {
                drop(entries);
                self.misses.fetch_add(1, Ordering::Relaxed);
                None
            }
        }
    }

    fn put(&self, identity_id: &str, org_id: &str, permissions: Vec<EffectivePermission>) {
        let key = PermissionCacheKey::new(identity_id, org_id);
        let entry = Entry {
            permissions,
            expires_at: Instant::now() + self.ttl,
        };
        self.entries
            .lock()
            .expect("cache lock poisoned")
            .insert(key, entry);
        self.puts.fetch_add(1, Ordering::Relaxed);
    }

    fn invalidate(&self, identity_id: &str, org_id: &str) {
        let key = PermissionCacheKey::new(identity_id, org_id);
        let removed = self
            .entries
            .lock()
            .expect("cache lock poisoned")
            .remove(&key)
            .is_some();
        if removed {
            self.targeted_invalidations.fetch_add(1, Ordering::Relaxed);
        }
    }

    fn invalidate_all(&self) {
        self.entries
            .lock()
            .expect("cache lock poisoned")
            .clear();
        self.wholesale_invalidations
            .fetch_add(1, Ordering::Relaxed);
    }
}

File: sdk-python/gatedhouse/_permission_cache.py
# Copyright (c) 2026 12vectors.com
# SPDX-License-Identifier: MIT
# See the LICENSE file in the repository root for the full license text.

"""Pluggable cache for the per-identity effective permission set."""

from __future__ import annotations

from abc import ABC, abstractmethod
from datetime import datetime, timedelta, timezone
from threading import Lock
from typing import Callable

from ._types import EffectivePermission, PermissionCacheKey


class PermissionCache(ABC):
    """Sits in front of the recursive-CTE permission-resolution query.

    The library ships an :class:`InMemoryPermissionCache` default. Hosts
    that need a shared cache (Redis, Memcached, etc.) implement this
    interface against their cache client and pass it to
    ``GatedhouseConfig(permission_cache=...)``.

    Implementations must be safe for concurrent use by multiple threads.
    """

    @abstractmethod
    def get(self, identity_id: str, org_id: str) -> list[EffectivePermission] | None:
        """Return the cached list for ``(identity_id, org_id)``, or
        ``None`` on miss / expiry."""

    @abstractmethod
    def put(self, identity_id: str, org_id: str,
            permissions: list[EffectivePermission]) -> None:
        """Cache the effective-permission list. Implementations should
        treat the supplied list as immutable (copy if needed)."""

    @abstractmethod
    def invalidate(self, identity_id: str, org_id: str) -> None:
        """Drop the cache entry for one identity in one org."""

    @abstractmethod
    def invalidate_all(self) -> None:
        """Drop every cached entry."""


class InMemoryPermissionCache(PermissionCache):
    """Default :class:`PermissionCache`: thread-safe, in-process,
    TTL-expiring dict. Lazy eviction on read.

    Exposes counters for hits, misses, puts, and invalidations as
    runtime observability — and to let tests prove caching behaviour,
    not just outcomes.
    """

    DEFAULT_TTL = timedelta(seconds=60)

    def __init__(
        self,
        ttl: timedelta = DEFAULT_TTL,
        clock: Callable[[], datetime] | None = None,
    ) -> None:
        self._ttl = ttl
        self._clock = clock or (lambda: datetime.now(timezone.utc))
        self._entries: dict[PermissionCacheKey, _Entry] = {}
        self._lock = Lock()
        self._hits = 0
        self._misses = 0
        self._puts = 0
        self._targeted_invalidations = 0
        self._wholesale_invalidations = 0

    # ---- PermissionCache --------------------------------------------------

    def get(self, identity_id: str, org_id: str) -> list[EffectivePermission] | None:
        key = PermissionCacheKey(identity_id, org_id)
        now = self._clock()
        with self._lock:
            entry = self._entries.get(key)
            if entry is None:
                self._misses += 1
                return None
            if now > entry.expires_at:
                self._entries.pop(key, None)
                self._misses += 1
                return None
            self._hits += 1
            return entry.permissions

    def put(self, identity_id: str, org_id: str,
            permissions: list[EffectivePermission]) -> None:
        key = PermissionCacheKey(identity_id, org_id)
        expires_at = self._clock() + self._ttl
        # Defensive copy so a later mutation by the caller can't poison
        # the cache.
        snapshot = list(permissions)
        with self._lock:
            self._entries[key] = _Entry(snapshot, expires_at)
            self._puts += 1

    def invalidate(self, identity_id: str, org_id: str) -> None:
        key = PermissionCacheKey(identity_id, org_id)
        with self._lock:
            if self._entries.pop(key, None) is not None:
                self._targeted_invalidations += 1

    def invalidate_all(self) -> None:
        with self._lock:
            self._entries.clear()
            self._wholesale_invalidations += 1

    # ---- observability ----------------------------------------------------

    def hit_count(self) -> int:
        with self._lock:
            return self._hits

    def miss_count(self) -> int:
        with self._lock:
            return self._misses

    def put_count(self) -> int:
        with self._lock:
            return self._puts

    def targeted_invalidation_count(self) -> int:
        with self._lock:
            return self._targeted_invalidations

    def wholesale_invalidation_count(self) -> int:
        with self._lock:
            return self._wholesale_invalidations

    def size(self) -> int:
        with self._lock:
            return len(self._entries)

    def reset_stats(self) -> None:
        """Reset all counters to zero. Does not touch cached entries."""
        with self._lock:
            self._hits = 0
            self._misses = 0
            self._puts = 0
            self._targeted_invalidations = 0
            self._wholesale_invalidations = 0


class _Entry:

    __slots__ = ("permissions", "expires_at")

    def __init__(self, permissions: list[EffectivePermission], expires_at: datetime) -> None:
        self.permissions = permissions
        self.expires_at = expires_at
